edX Insights installation guide

mulby opened the pull request:
Very early draft of the high level deployment documentation. This will likely require significant editing, so don't be shy!

Target audience is system administrators who are looking to install the analytics stack.

I used a tool to put in hard line breaks at 80 characters. Not sure what the convention is with that. I have a bunch of questions, but hopefully we can iron all of that out in this PR.

Reviewers: @lamagnifica @e0d

FYI: @brianhw @rocha @clintonb

lamagnifica commented on en_us/install_operations/source/analytics/install_analytics.rst:
```diff
@@ -0,0 +1,212 @@
+See the following sections:
+
+* `Overview`_
+
```
xref each of your 2nd level headings: What You Should Know Before You Start, Planning Your Deployment, etc.

mulby replied:
Done.

lamagnifica commented on en_us/install_operations/source/analytics/install_analytics.rst:
```diff
@@ -0,0 +1,212 @@
+Overview
+********
+
+edX Insights provides Course Staff access to data gathered from courses that
```
We capitalize "EdX" when it starts a sentence. Tena apologizes and agrees that we shouldn't but that's how she set up the branding
"provides Course Staff access" is a little confusing as written, as though maybe permissions are being granted. 
maybe, 
Course teams use edX Insights to access data gathered... (?)

mulby replied:
Done (hopefully I found all of them)

lamagnifica commented on en_us/install_operations/source/analytics/install_analytics.rst:
```diff
@@ -0,0 +1,212 @@
+are being used by students. edX Insights presents this information to the
+Course Staff via charts, summary statistics, and data tables.
+
+The LMS gathers data about student activity. This data is aggregated by the edX
```
For this first use of LMS, spell out "Learning Management System (LMS)". In the rest of the file you can just use "LMS".

mulby replied:
Done

lamagnifica commented on en_us/install_operations/source/analytics/install_analytics.rst:
```diff
@@ -0,0 +1,212 @@
+******************************
+
+The Learning Management System records student actions in tracking log files.
+These files are periodically compressed and copied into a filesystem that can be
```
what does the compress & copy action? in the diagram, it looks like the pipeline does it. Worth explicitly stating

mulby replied:
It's actually yet another tool called "logrotate". It is a semi-standard linux system administration tool.

I've updated this text though.

lamagnifica commented on en_us/install_operations/source/analytics/install_analytics.rst:
```diff
@@ -0,0 +1,212 @@
+
+The edX Analytics Pipeline reads the MySQL database used by the LMS as well as
+the tracking log files produced by the LMS. The data is processed and the
+resulting summary data is published to the *Result Store*. The result store is a
```
no need to capitalize result store

mulby replied:
fixed capitalization

Also, I've re worked the diagram to more closely match the text. Let me know if it still isn't close enough... I'm trying to walk the line between too little and too much detail.

lamagnifica commented on en_us/install_operations/source/analytics/install_analytics.rst:
```diff
@@ -0,0 +1,212 @@
+The Learning Management System records student actions in tracking log files.
+These files are periodically compressed and copied into a filesystem that can be
+read by the edX Analytics Pipeline. The Learning Management System also captures
+a lot of information in a MySQL database. The edX Analytics Pipeline connects
```
should the MySQL database be represented inside the LMS box in the diagram?

mulby replied:
Done.

lamagnifica commented on en_us/install_operations/source/analytics/install_analytics.rst:
```diff
@@ -0,0 +1,212 @@
+The Scheduler schedules the execution of data computation tasks. Data
+computation tasks are run by the edX Analytics Pipeline. Data computation tasks
+are used to update parts of the Result Store.
+
```
referring to the diagram again, is the scheduler related to the batch process? is the data computation described here completed by the batch process?
I'd suggest making what you show in the diagram relate more directly to what you describe in text.

mulby replied:
Done.

lamagnifica commented on en_us/install_operations/source/analytics/install_analytics.rst:
```diff
@@ -0,0 +1,212 @@
+edX Insights uses the edX Analytics Data API to present data to users. Users
+access the data using a supported web browser. edX Insights communicates
+directly with the Learning Management System to authenticate users, authorize
+users and read course structure information.
```
the direct path between Insights and the LMS is not shown in the diagram. is there really a direct connection? Maybe there should be an arrow coming out of the Pipeline box and pointing to the left, back at the LMS box...

mulby replied:
There is, but I'm nervous it would make the diagram to complex... maybe swing by and we can play with it a little bit to see if we can make it work? It is shown in the more detailed diagram below...

lamagnifica commented on en_us/install_operations/source/analytics/install_analytics.rst:
```diff
@@ -0,0 +1,212 @@
+
+Amazon's Elastic MapReduce service offers pre-configured Hadoop clusters. If you
+are able to use Amazon Web Services, it is recommended you use this service.
+Proper installation and configuration of Hadoop can be time consuming.
```
it is recommended you use this service --> use of this service is recommended (?)

mulby replied:
Done

lamagnifica commented on en_us/install_operations/source/analytics/install_analytics.rst:
```diff
@@ -0,0 +1,212 @@
+Additionally, vendors such as Cloudera and MapR offer simplified Hadoop
+administration experiences.
+
+Hadoop clusters can be scaled vertically and horizontally as your data grows.
```
Is this sentence in the right place?
maybe is could move to the end, after the next paragraph? or possibly earlier, after the sentence that ends on line 124?

mulby replied:
Done

lamagnifica commented on en_us/install_operations/source/analytics/install_analytics.rst:
```diff
@@ -0,0 +1,212 @@
+edX Applications
+================
+
+The edX Analytics Data API will be responding to a small number of requests
```
will be responding --> responds

lamagnifica commented on en_us/install_operations/source/analytics/install_analytics.rst:
```diff
@@ -0,0 +1,212 @@
+============
+
+The results of computations performed by the edX Analytics Pipeline are stored
+in a MySQL database. Even small installations are recommended to use a different
```
is this db in addition to the one that the LMS populates, or the same? (if the same, it should be described in the same way each time)

lamagnifica commented on en_us/install_operations/source/analytics/install_analytics.rst:
```diff
@@ -0,0 +1,212 @@
+Database Service is used to deploy a MySQL server into a private subnet. The
+MySQL database will be used as the Result Store.
+
+.. image:: /Images/Analytics_AWS_Deployment.png
```
I'd suggest moving the image up to line 190, after the first sentence.

mulby replied:
Done.

lamagnifica commented on en_us/install_operations/source/analytics/install_analytics.rst:
```diff
@@ -0,0 +1,212 @@
+============
+
+.. image:: /Images/Analytics_Pipeline.png
+
```
for accessibility, we provide an "alt" for every image. it should briefly describe the purpose of the image...

mulby replied:
Done.

lamagnifica commented on en_us/install_operations/source/analytics/install_analytics.rst:
```diff
@@ -0,0 +1,212 @@
+good candidate.
+
+===================
+Example Deployments
```
I'd make this a 2nd level heading

mulby replied:
Done.

lamagnifica commented on en_us/install_operations/source/analytics/install_analytics.rst:
```diff
@@ -0,0 +1,212 @@
+What You Should Know Before You Start
+*************************************
+
+This guide assumes that you:
```
global English tip: it's hard to translate sentence fragments. Suggest making this introduction to the list a complete sentence. 
maybe,
You must understand the following concepts to install edX Insights and deploy the edX Analytics pipeline.

lamagnifica commented:
Gabe, I think this is great. Thanks!

mulby commented:
@rocha this is the doc PR we discussed during standup

@lamagnifica here is the second draft

lamagnifica commented:
:+1: Thanks for incorporating my suggestions!

mhoeber commented on en_us/install_operations/source/analytics/install_analytics.rst:
```diff
@@ -0,0 +1,223 @@
+********
+
+Course teams use edX Insights to access to data gathered from courses that are
+being used by students. Course teams use edX insights to display charts, summary
```
cap Insights

mhoeber commented on en_us/install_operations/source/analytics/install_analytics.rst:
```diff
@@ -0,0 +1,223 @@
+The Learning Management System (LMS) gathers data about student activity. This
+data is aggregated by the edX Analytics Pipeline. The aggregated data is exposed
+by the edX Analytics Data API. EdX Insights reads the data from the edX
+Analytics Data API and presents the data to course team members.
```
could add a link to api doc at:  http://edx.readthedocs.org/projects/edx-data-analytics-api/en/latest/index.html

mhoeber commented on en_us/install_operations/source/analytics/install_analytics.rst:
```diff
@@ -0,0 +1,223 @@
+============
+
+The results of computations performed by the edX Analytics Pipeline are stored
+in a MySQL database. Even small installations are recommended to use a different
```
reword as:  Even small installations should use a different MySQL server than the one used by the LMS.

mhoeber commented:
@mulby  this is great, thank you.  I added a few small comments; can you also add a change log entry with link to the new page?

mulby commented:
@mhoeber thanks for the feedback, I've addressed your comments. While I was at it I added some more links to external utilities that I reference in the doc. Hopefully I didn't go overboard...

@e0d can you review this when you get a chance?

@rocha or @brianhw can you also take a look?

mhoeber commented:
:+1:

rocha commented:
Looks good to me :rocket:

Should authentication/authorization be included in this document too? At least mentioned as a post-deployment step? 

We may also want to mention that the pipeline copies data from the read-replica using scoop. This implies a certain type of access pattern which may help deployers determine what systems they should use.

rocha commented:
@mulby also, change the tittle of the PR before merging.

e0d commented:
Can't comment on images too easily.  I don't understand the peering between insights and the LMS, is that meant to represent the OAuth conversation?

mulby commented:
@e0d it is supposed to represent a secure connection that is enabled by VPC peering. It actually represents more than just OAuth, since we will soon be using it for other API calls (course metadata etc).

I tried to represent the connection from the pipeline to the LMS DB in a similar way.

e0d commented on en_us/install_operations/source/analytics/install_analytics.rst:
```diff
@@ -0,0 +1,230 @@
+edX Analytics Data API
+**********************
+
+The edX Analytics Data API provides an HTTP interface for accessing data in the
```
nit whitespace

mulby replied:
heh, what whitespace are you referring to?

e0d commented:
This doc seems to be more about high-level architecture, best practices and running hadoop/EMR rather than installing insights.  The document is very useful, but I think we should either add some sample plays for installing analytics-api and insights or rename.

mulby commented:
@e0d An excellent point, we have follow-on work to update this document as we develop an automated process. We don't yet know what that process is going to look like (cloud migrations, plays etc).

@lamagnifica and @mhoeber what do you think? should we hold off on merging this until we have more detail about the actual procedure? merge it but don't include it in the index? rename it for the time being?

lamagnifica commented:
I say we publish it, named as is. In the introduction, we might want to add words to the effect that "Work to prepare complete installation procedures for edX Insights is in progress. Introductory material is available now."

mulby commented:
@e0d I added a disclaimer to the introduction as @lamagnifica suggested

e0d commented:
:+1:
